### PR TITLE
[WIP] Add more documentation for entities

### PR DESCRIPTION
**Pull request by jyn514**
Addresses https://github.com/CraneStation/cranelift/issues/1038

Document stackslots, funcrefs, heaps, sigrefs, and jumptables.
Also adds more inter-crate links.

## Undocumented

- [x] `Constant` - https://github.com/CraneStation/cranelift/pull/1041#issuecomment-532348093
- [x] `GlobalValue` - https://github.com/CraneStation/cranelift/pull/1041#issuecomment-532575257
- [x] `Table` - https://github.com/CraneStation/cranelift/pull/1041#issuecomment-532575257
- [x] `Inst` - https://github.com/CraneStation/cranelift/pull/1041#issuecomment-532575257

## Issues

I can't find a way to link to a type/function in a different crate if that crate isn't currently in scope for the code. For example, [the documentation for JumpTable](https://github.com/CraneStation/cranelift/compare/master...jyn514:doc-entity?expand=1#diff-dd78b3709a42d640a60ea2733dcdbe53R187) links to `FunctionBuilder::create_jump_table`, but the link is broken (it points to `cranelift_codegen/ir/entities/cranelift_frontend::FunctionBuilder::create_jump_table`, which is clearly wrong).

There are two ways I can think to resolve this:

1. Link directly to docs.rs like this: `https://docs.rs/cranelift-module/*/cranelift_module/struct.Module.html`. This has the downside that it will always point to the latest version (and won't work locally or offline).
2. Put the name of the type in prose and have users search for it. This is not ideal but will work the same everywhere.

### CODE REVIEW

**Comment by abrown** [comment]
@jyn514 I can explain Constant: it is an integer used to index into the ConstantPool to identify constant values held there (see https://github.com/CraneStation/cranelift/blob/dcaa7f5ba618d72c9dd24a29c9a485dc24ab0d07/cranelift-codegen/src/ir/constant.rs#L97).

**Comment by jyn514** [comment]
I see, and it's public because `cranelift-frontend` needs to access it?

**Comment by abrown** [comment]
> I can't find a way to link to a type/function in a different crate if that crate isn't currently in scope for the code

Also, I believe I had similar issues trying to fix some documentation warnings; in the end I went with something like `crate::ir::...` and that seemed to work: https://github.com/CraneStation/cranelift/pull/952. The trick is to use correct Rust paths (see links in https://stackoverflow.com/a/53504254 and https://github.com/rust-lang/rust/issues/43466). I think that getting the Rust paths to work is better long-term than direct links or prose.

File: cranelift-codegen/src/ir/entities.rs [hunk 1]
```diff
@@ -44,6 +48,18 @@ impl Ebb {
...
+///
+/// You can get a constant `Value` from the following
+/// [`InstBuilder`](super::InstBuilder) instructions:
+///
+/// - [`iconst`](super::InstBuilder::iconst) for integer constants
+/// - [`f32const`](super::InstBuilder::f32const) for 32-bit float constants
+/// - [`f64const`](super::InstBuilder::f64const) for 64-bit float constants
+/// - [`bconst`](super::InstBuilder::bconst) for boolean constants
+/// - [`vconst`](super::InstBuilder::vconst) for vector constants
+/// - [`null`](super::InstBuilder::null) for null pointer constants
```
Comment (bjorn3) [inline cranelift-codegen/src/ir/entities.rs]: *null reference

File: cranelift-codegen/src/ir/entities.rs [hunk 2]
```diff
@@ -68,6 +84,16 @@ pub struct Inst(u32);
...
+///
+/// Stack slots represent an address on the
+/// [call stack](https://en.wikipedia.org/wiki/Call_stack).
+///
+/// `StackSlot`s can be created with
+/// `cranelift_frontend::FunctionBuilder::create_stackslot`.
+///
+/// `StackSlot`s are most often used with
+/// [`stack_addr`](super::InstBuilder::stack_addr) and
+/// [`stack_load`](super::InstBuilder::stack_load).
```
Comment (bjorn3) [inline cranelift-codegen/src/ir/entities.rs]: missing stack_store

**Comment by jyn514** [comment]
> in the end I went with something like crate::ir::...

That only works if the crate is imported into the current scope, though. I can't put `extern crate cranelift_module;` in the code because that breaks the code of anyone who uses cranelift for WASM.

File: cranelift-codegen/src/ir/entities.rs [hunk 3]
```diff
@@ -122,7 +148,13 @@ impl Constant {
...
 }
 
-/// An opaque reference to a jump table.
+/// An opaque reference to a [jump table](https://en.wikipedia.org/wiki/Branch_table).
+///
+/// `JumpTable`s are used for indirect branching.
+/// See [`br_table`](super::InstBuilder::br_table) for details.
+///
+/// `JumpTable`s can be created with
+/// [`create_jump_table`](cranelift_frontend::FunctionBuilder::create_jump_table).
```
Comment (bjorn3) [inline cranelift-codegen/src/ir/entities.rs]: If you want a jump table which doesnt start at 0, or is not contiguous, you should use cranelift_frontend::Switch instead.

**Comment by abrown** [comment]
> I see, and it's public because cranelift-frontend needs to access it?

Now that I look at its usages, I don't see it anywhere except in codegen so it might be able to be downgraded to `pub(crate)`. The idea, though, is that it is part of a public API for interacting with the constant pool.

File: cranelift-codegen/src/ir/entities.rs [hunk 4]
```diff
@@ -159,7 +206,17 @@ impl FuncRef {
...
+///
+/// `SigRef`s are used to declare a function with
+/// [`import_function`](cranelift_frontend::FunctionBuilder::import_function)
+/// as well as to make an [indirect function call](super::InstBuilder::call_indirect).
+///
+/// `SigRef`s can be created with
+/// [`import_signature`](cranelift_frontend::FunctionBuilder::import_signature).
+///
+/// You can retrieve the [`Signature`](super::Signature) that was used to create a `SigRef` with
+/// [`signature`](cranelift_frontend::FunctionBuilder::signature).
```
Comment (bjorn3) [inline cranelift-codegen/src/ir/entities.rs]: func.dfg.signatures

**Comment by abrown** [comment]
Maybe https://github.com/rust-lang/rust/issues/43466#issuecomment-500418346 is fixed somewhere? Not exactly sure...

**Comment by abrown** [comment]
> Maybe rust-lang/rust#43466 (comment) is fixed somewhere? Not exactly sure...

But if there is absolutely no way to link to external crates then I guess my vote would be for the direct link.

**Comment by jyn514** [comment]
> The idea, though, is that it is part of a public API for interacting with the constant pool.

I don't see a way to get a `Constant` without inserting things into the pool yourself, though ... I guess it could be useful as a data structure so you aren't caching things twice (once in cranelift and once in the frontend).

**Comment by abrown** [comment]
@jyn514, for `Inst` you might want to look at how it is used in `DataFlowGraph`.

**Comment by bnjbvr** [comment]
Thanks for taking a stab at this!

Note that all of these implement EntityRef, so they're all handles to some data living somewhere else. For instance Constant has ConstantData, etc. I remember proposing to constantly suffix them with "Ref" or something, but this was a long time ago and I'm unsure what the outcome of the discussion was.

- GlobalValue can be mapped to a wasm global value, but doesn't need to; it is "just" a value that may be live across the entire function lifetime, in my understanding, and that can be preloaded from other global values. (Useful when you have a VmContext, and want to load constant things from it, for instance)
- Table is used for wasm function tables at the moment. I don't know if it's planned to use it for other table kinds in the future, could be. Useful for implementing indirect calls to known functions.
- Inst is the handle for an instruction. I imagine most usage is internal, and I wonder if this could be made not public.

**Comment by bjorn3** [comment]
> Inst is the handle for an instruction. I imagine most usage is internal, and I wonder if this could be made not public.

I do use it to attach a comment to a specific instruction in the clif ir written by cg_clif.

**Comment by jyn514** [comment]
Getting slightly off-topic now - is there any way that [`pretty_clif.rs`](https://github.com/bjorn3/rustc_codegen_cranelift/blob/master/src/pretty_clif.rs) could be made part of the main Cranelift API? I would love to reuse that code without having to copy/paste most of the implementation.

**Comment by jyn514** [comment]
> Useful when you have a VmContext, and want to load constant things from it

Is this a WASM thing? It mentions a sandbox in [the documentation](https://docs.rs/cranelift-codegen/0.42.0/cranelift_codegen/ir/enum.ArgumentPurpose.html#variant.VMContext), and I don't see a way to construct it from `cranelift_frontend` or `cranelift_codegen`

**Comment by bjorn3** [comment]
> is there any way that pretty_clif.rs could be made part of the main Cranelift API?

https://github.com/CraneStation/cranelift/pull/379#issuecomment-401552747

> I would love to reuse that code without having to copy/paste most of the implementation.

Please note that currently comments get inserted **after** the specified inst, as there is no way to know what the next inst id will be.

**Comment by jyn514** [comment]
This is ready for review as soon as I fix the broken links. @abrown recommended permalinks to docs.rs, should I do that?

File: cranelift-codegen/src/ir/entities.rs [hunk 5]
```diff
@@ -78,7 +78,18 @@ impl Value {
...
+/// Most usage of `Inst` is internal. `Inst`ructions are returned by
+/// [`InstBuilder`](super::InstBuilder) instructions that do not return a
+/// [`Value`], such as control flow and trap instructions.
+///
+/// If you look around the API, you can find many inventive uses for `Inst`,
+/// such as [annotating specific instructions with a comment][inst_comment]
+/// or [performing reflection at compile time](super::DataFlowGraph::analyze_branch)
+/// on the type of instruction.
+///
+/// [inst_comment]: https://github.com/CraneStation/cranelift/pull/1041#issuecomment-532575989
```
Comment (bjorn3) [inline cranelift-codegen/src/ir/entities.rs]: Maybe link https://github.com/bjorn3/rustc_codegen_cranelift/blob/0f8814fd6da3d436a90549d4bb19b94034f2b19c/src/pretty_clif.rs instead.

**Comment by abrown** [comment]
@jyn514 I'm fine merging this but we need to fix the build issues: https://dev.azure.com/CraneStation/Cranelift/_build/results?buildId=897&view=logs.

**Comment by jyn514** [comment]
Ok, I'll change it to link directly to docs.rs, then.

**Comment by jyn514** [comment]
This is ready for review.

**Comment by abrown** [comment]
@jyn514 thanks for the fix!

**Pull request merged**